RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+
+
+### Forwards Compatibility
+
+Forwards compatibility (BC), the ability for older versions of PyTorch to run programs from newer versions, is important so users don’t need to update PyTorch.
+
+PyTorch makes the following FC promises today:
+
+
+
+* **OSS** — no promise
+* **FB Internal** — PyTorch commits can run existing PyTorch eager, package/deploy, and serialized torchscript programs for at least two weeks
+    * The addition of a new kwarg-only argument at the end of an op’s parameter list (but before out=, if present) with a default value is FC-compatible for serialized [torchscript](https://fb.workplace.com/groups/pytorch.dev/permalink/909079013003913/) and [edge](https://fb.workplace.com/groups/pytorch.dev/permalink/912379562673858/).
+
+
+## Goals
+
+
+
+* To support backward and (some) forward compatibility for an arbitrary BC or FC break update (schema, functional, etc) on an operator by [versioning](https://docs.google.com/document/d/1nyXmss2O003ZgKrhDmd-kyLNjjMqEXww_2skOXqkks4/edit). 
+* It’s a systematic flow to prevent BC/FC breakage on both deploy and runtime stages.
+* To provide testing that accurately detects dangerous BC and FC-breaking changes.
+* To support both server and edge use cases, including TorchScript, package/deploy and Edge.
+
+
+## Non-goals
+
+
+
+* It does not mean that models with old operator schema can **always** run successfully on new runtime and vice versa. 
+* It’s not for the “automatic” BC/FC support that can be done without any developer’s manual work (for example, the number of arguments mentioned in the Context). To apply versioning on the updated operator, the author needs to manually add a line in the version table and provide the resolution function for BC. This proposal is for BC/FC breakages that the automatic supports don’t apply. 
+* It does not include the BC/FC for package/deploy itself. The Python-only operators are transparent to TS and Edge clients, with the TS compilation. 
+
+
+# Proposal
+
+
+
+* **Eager changes**
+    * operator_versions.yaml** and **operator_upgraders.py** are added to register operator upgrades that are BC/FC breaking. **
+
+<p id="gdcalert1" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: undefined internal link (link text: "Example changes"). Did you generate a TOC? </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert2">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+
+[Example changes](#heading=h.enajdjh00ukn)
+        * Note: this will not cover functional operators
+        * The default value is zero
+    * **Use a single operator version number for all operators**
+        * This number may be shared by the deploy version, but separate from other file format versions
+    * **Older versions of operators registration are kept but they should only be matchable with special circumstances**
```

Comment:
This is a little vague -- can we be more precise here, maybe with an example?

Reply:
What are these special circumstances?

Reply:
I don't think it's the case anymore, if the upgraders are provided. Will remove this item and we could revisit it later.